File: YTTabBarControler(OC)/YTTabBar.m
//
//  YTTabBar.m
//  02-自定制TabBar
//
//  Created by 余婷 on 16/6/4.
//  Copyright © 2016年 余婷. All rights reserved.
//

#import "YTTabBar.h"
#import "YTTabBarButton.h"

#define Btn_tag 100
#define Center_tag 1000

@interface YTTabBar(){

    BOOL _haveBackgroundImage;
}



@end

@implementation YTTabBar

- (void)setBackgroundImage:(UIImage *)backgroundImage{

    UIImageView * imageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, 49)];
    imageView.image = backgroundImage;
    imageView.contentMode = UIViewContentModeBottom;
    [self addSubview:imageView];
    [self sendSubviewToBack:imageView];
    
    _haveBackgroundImage = YES;
    
}

#pragma mark - 创建每个控制器对应的按钮
- (void)addbuttonWithItem:(UITabBarItem *)item{

    //1.创建一个按钮,不需要设置frame
    YTTabBarButton * button = [[YTTabBarButton alloc] init];
    //2.设置item属性
    button.item = item;
    //3.显示在界面上
    [self addSubview:button];
    //4.添加事件
    [button addTarget:self action:@selector(buttonOnclicked:)];
}

#pragma mark - 按钮点击
- (void)buttonOnclicked:(YTTabBarButton *)button{
    
    //把其他按钮变成非选中状态
    //获取到原来选中的按钮
    YTTabBarButton * tButton = [self viewWithTag:self.selectedIndex + Btn_tag];
    tButton.isSelected = NO;
    tButton.userInteractionEnabled = YES;
    

    //把当前点击的按钮变成选中状态
    button.isSelected = YES;
    self.selectedIndex = button.tag - Btn_tag;
    button.userInteractionEnabled = NO;
    
    
    
}

#pragma mark  - 改变选中下标
- (void)setSelectedIndex:(NSInteger)selectedIndex{

    _selectedIndex = selectedIndex;
    //让tabBarController去切换视图控制器
    //YTTaBar想要切换视图控制器，但是它做不到，需要tabBarController去帮他做这件事。
    //1.确定三要素
    //协议:切换到指定的视图控制器
    //代理:tabBarController
    //委托:YTTaBar
    [self.delegate changeControllerWithIndex:self.selectedIndex];
    
}


#pragma mark - 设置每个按钮的frame
- (void)layoutSubviews{

    //计算每个按钮的frame
    [self setFrameWithCenter];
    
}

//当考虑有centerView的时候
- (void)setFrameWithCenter{

    //1.通用属性
    CGFloat W = self.frame.size.width;
    CGFloat H = self.frame.size.height;
    
    //按钮的个数
    NSInteger buttonCount = self.subviews.count;
    if (_haveBackgroundImage) {
        
        buttonCount -= 1;
    }
    //有centerView的情况
    //centerView的相关属性
    CGFloat centerW = self.centerView.frame.size.width;
    CGFloat centerH = self.centerView.frame.size.height;
    CGFloat centerX = (W - centerW)/ 2.0f;
    CGFloat centerY;
    if (centerH <= H) {
        centerY = H / 2.0f - centerH / 2.0f;
    }else{
        
        centerY = H - centerH;
    }
    self.centerView.frame = CGRectMake(centerX, centerY, centerW, centerH);
    
    if (self.centerView) {
        
        buttonCount  -= 1;
    }else{
    
        centerW = 0;
        centerH = 0;
        
    }
    
    //按钮的frame相关
    CGFloat buttonY = 0;
    CGFloat buttonH = self.frame.size.height - 5;
    CGFloat buttonW = (W - centerW) / buttonCount;
    CGFloat buttonX;
    
    int  i = 0;
    for (UIView * tView in self.subviews) {
        //拿到按钮
        if ([tView isKindOfClass:[YTTabBarButton class]]) {
            
            YTTabBarButton * button = (YTTabBarButton *)tView;
            
            //计算frame
            //前面部分
            if (i < buttonCount/2) {
                
                buttonX = i * buttonW;
                
            //后面部分
            }else{
            
                buttonX = i * buttonW + centerW;
            }
            
            //设置frmame
            button.frame = CGRectMake(buttonX, buttonY, buttonW, buttonH);
            
            //设置默认选中的按钮
            if (i == self.selectedIndex) {
                
                button.isSelected = YES;
            }
            
            //设置tag值
            button.tag = Btn_tag + i;
            
            
            i ++;
        }
        
    }
    
    
}

- (void)setButtonFrame{

    //1.通用属性
    //按钮的个数
    NSInteger buttonCount = self.subviews.count;
    CGFloat buttonY = 0;
    CGFloat buttonH = self.frame.size.height;
    CGFloat buttonW = self.frame.size.width / buttonCount;
    
    //2.遍历拿到tabBar上所有的按钮
    int i = 0;
    for (YTTabBarButton * button  in self.subviews) {
        CGFloat buttonX = i * buttonW;
        //设置frame
        button.frame = CGRectMake(buttonX, buttonY, buttonW, buttonH);
        //设置tag值
        button.tag = i + Btn_tag;
        
        //设置默认选中的按钮
        if (i == self.selectedIndex) {
            
            button.isSelected = YES;
        }
        
        i++;
    }
}

#pragma mark -centerView赋值
- (void)setCenterView:(UIView *)centerView{

    _centerView = centerView;
    
    //显示在界面上
    UIView * view = [self viewWithTag:Center_tag];
    if (view) {
        [view removeFromSuperview];
    }else{
        _centerView.tag = Center_tag;
        [self addSubview:_centerView];
        
    }
    
}

@end

File: YTTabBarControllerDemo(Swift)/YTTabBarControllerDemo/YTTabBarController/YTTabBar.swift
//
//  YTTabBar.swift
//  01-自定制tabBarController
//
//  Created by 余婷 on 16/9/10.
//  Copyright © 2016年 余婷. All rights reserved.
//

import UIKit

class YTTabBar: UIView {

    
    //MARK: - 属性
    ///1.当前选中的下标
    var selectedIndex = 1
    
    ///2.声明闭包用来传值
    var changeSelectedIndex:((Int)->Void)? = nil
    
    ///3.设置按钮的文字颜色
    var titleColor = UIColor.blackColor()
    
}

//MARK: - 添加按钮
extension YTTabBar{

    ///添加按钮
    func addButtonWithItem(item:UITabBarItem) {
        
        //创建按钮
        let btn = YTTabBarButton(tabBarItem: item)
        //添加点击事件
        btn.addTarget(self, action: "btnAction:")
        
        //设置按钮的tag值
        btn.tag = 100+self.subviews.count
        
//        //判断是否是需要选中按钮
//        if self.selectedIndex == self.subviews.count{
//            
//            btn.isSelected = true
//        }
        
        //添加到界面上
        self.addSubview(btn)
    }
}

//MARK: - 按钮点击
extension YTTabBar{

    func btnAction(btn:YTTabBarButton){
        
        //将原来处于选中状态的按钮变成非选中状态
        let selectBtn = self.viewWithTag(100+self.selectedIndex) as! YTTabBarButton
        selectBtn.isSelected = false
        
        //将当前按下的按钮变成选中状态
        btn.isSelected = true
        //更新选中下标
        self.selectedIndex = btn.tag - 100
        
        //通知tabBarController切换视图控制器
        self.changeSelectedIndex!(self.selectedIndex)
        
    }
}

//MARK: - 计算子视图的frame
extension YTTabBar{

    override func layoutSubviews() {
        super.layoutSubviews()
        
        //通用
        let btnW = self.frame.size.width / CGFloat(self.subviews.count)
        let btnH = self.frame.size.height
        let btnY:CGFloat = 0
        
        //遍历拿到所有的按钮
        for (i,item) in self.subviews.enumerate() {
            
            let btn = item as! YTTabBarButton
            
            let btnX = CGFloat(i) * btnW
            //1.设置frame
            item.frame = CGRectMake(btnX, btnY, btnW, btnH)
            
            //2.设置默认选中的按钮
            if i == self.selectedIndex {
                
                btn.isSelected = true
            }
            //3.设置按钮颜色
            btn.titleLabel.textColor = self.titleColor
            
        }
        
        
        
    }
}
